File: src/LongestSubstringWithoutRepeatingCharacters.java
import java.util.HashSet;
import java.util.Set;

public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        int l = s.length();
        if (l <= 1) {
            return l;
        }
        Set<Character> charSet = new HashSet<>();
        int i = 0;
        int max = 0;
        while (i < l) {
            int j = i + charSet.size();
            while (j < l && !charSet.contains(s.charAt(j))) {
                charSet.add(s.charAt(j));
                j++;
            }
            int newL = j - i;
            if (newL > max) {
                max = newL;
            }
            charSet.remove(s.charAt(i));
            i++;
        }
        return max;
    }

    public static void main(String[] args) {
        LongestSubstringWithoutRepeatingCharacters l = new LongestSubstringWithoutRepeatingCharacters();
        System.out.println(l.lengthOfLongestSubstring("dvdf"));
    }
}

File: Python/Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        if n == 0:
            return 0
        char_set = set()
        left, right = 0, 0
        res = 0
        while left < n and right < n:
            if s[right] in char_set:
                char_set.remove(s[left])
                left += 1
            else:
                char_set.add(s[right])
                right += 1
                res = max(res, right - left)
        return res
